File: src/main/java/org/example/coin_laundry_app_backend/geo/external/NaverReverseGeocodingAdapter.java
package org.example.coin_laundry_app_backend.geo.external;

import org.example.coin_laundry_app_backend.geo.domain.model.EPSG4326Coordinate;
import org.example.coin_laundry_app_backend.geo.service.ReverseGeocodingService;
import org.example.coin_laundry_app_backend.geo.service.model.ReverseGeoModel;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Component
public class NaverReverseGeocodingAdapter implements ReverseGeocodingService {

    NaverReverseGeocodingAdapter(
        @Value("${naver.api.id}") final String naverApiKeyId,
        @Value("${naver.api.key}") final String naverApiKey
    ) {
        this.webClient = WebClient.builder()
            .baseUrl("https://naveropenapi.apigw.ntruss.com/map-reversegeocode/v2/gc")
            .defaultHeaders((headers) -> {
                    headers.add("X-NCP-APIGW-API-KEY-ID", naverApiKeyId);
                    headers.add("X-NCP-APIGW-API-KEY", naverApiKey);
                }
            )
            .build();
    }

    private final WebClient webClient;

    @Override
    public Mono<ReverseGeoModel> getReverseGeocoding(final EPSG4326Coordinate coordinate) {
        return webClient.get()
            .uri(uriBuilder -> uriBuilder
                .queryParam("coords", EPSG4326CoordinateBuilderForApi.build(coordinate))
                .queryParam("output", "json")
                .queryParam("orders", "addr")
                .queryParam("sourcecrs", "EPSG:4326")
                .build()
            )
            .retrieve()
            .bodyToMono(GetReverseGeocodingResponse.class)
            .map(response -> {
                var address = response.results().get(0).region();
                return new ReverseGeoModel(
                    address.area0().name(),
                    address.area1().name(),
                    address.area2().name(),
                    address.area3().name()
                );
            });
    }
}

File: carry-geo/src/main/kotlin/com/carry/geo/adapter/outbound/external/naver/NaverReverseGeocodingAdapter.kt
package com.carry.geo.adapter.outbound.external.naver

import com.carry.geo.adapter.outbound.external.naver.dto.NaverReverseGeocodingResponse
import com.carry.geo.application.port.outbound.ReverseGeocodingPort
import com.carry.geo.domain.exception.ReverseGeocodingFailedException
import com.carry.geo.domain.model.ReverseGeocodingResult
import com.carry.geo.domain.vo.Coordinate
import org.springframework.web.client.RestClient

/**
 * Naver 역지오코딩 API 호출 — Circuit Breaker + Redis 캐싱 데코레이터로 감싸 노출되므로
 * @Component가 아닌 [com.carry.geo.adapter.outbound.config.GeocodingResilienceConfig]에서
 * 명시적으로 빈 등록된다.
 */
class NaverReverseGeocodingAdapter(
    properties: NaverApiProperties,
) : ReverseGeocodingPort {

    private val restClient: RestClient = RestClient.builder()
        .baseUrl(properties.reverseGeocodingUrl)
        .defaultHeader("X-NCP-APIGW-API-KEY-ID", properties.clientId)
        .defaultHeader("X-NCP-APIGW-API-KEY", properties.clientSecret)
        .build()

    override fun reverseGeocode(coordinate: Coordinate): ReverseGeocodingResult {
        val response = try {
            restClient.get()
                .uri {
                    it.queryParam("coords", coordinate.toNaverCoordsFormat())
                        .queryParam("output", "json")
                        .queryParam("orders", "addr")
                        .queryParam("sourcecrs", "EPSG:4326")
                        .build()
                }
                .retrieve()
                .body(NaverReverseGeocodingResponse::class.java)
        } catch (e: Exception) {
            throw ReverseGeocodingFailedException(
                "네이버 역지오코딩 API 호출에 실패했습니다: ${e.message}", e,
            )
        }

        val result = response?.results?.firstOrNull()
            ?: throw ReverseGeocodingFailedException("역지오코딩 결과가 없습니다")

        return result.region.toDomain()
    }
}
